### PR TITLE
fix(regex-combination): share regex match groups for matchStrings

### PR DESCRIPTION
**Pull request by secustor**
## Changes:
Use regex matching groups of all `matchStrings` instead of the current one for templating, when using the "combination" `matchStringStrategy`.  

## Context:
https://github.com/renovatebot/renovate/discussions/11234

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/regex/index.spec.ts [hunk 1]
```diff
@@ -220,6 +220,24 @@ describe(getName(), () => {
     expect(res).toMatchSnapshot();
     expect(res.deps).toHaveLength(1);
   });
+  it('extracts with combination strategy and templates', async () => {
```
Comment (viceice) [inline lib/manager/regex/index.spec.ts:223]:
```suggestion

  it('extracts with combination strategy and templates', async () => {
```

Readability.

File: lib/manager/regex/index.spec.ts [hunk 2]
```diff
@@ -220,6 +220,24 @@ describe(getName(), () => {
...
+      depNameTemplate: 'helm_repo/{{{ depName }}}',
+    };
+    const res = await extractPackageFile(
+      exampleGitlabCiYml,
+      '.gitlab-ci.yml',
+      config
+    );
+    expect(res).toMatchSnapshot();
+    expect(res.deps).toHaveLength(1);
+  });
```
Comment (viceice) [inline lib/manager/regex/index.spec.ts:240]:
```suggestion
  });

```

File: lib/manager/regex/index.ts [hunk 3]
```diff
@@ -40,6 +40,7 @@ function regexMatchAll(regex: RegExp, content: string): RegExpMatchArray[] {
 
 function createDependency(
   matchResult: RegExpMatchArray,
+  combinedGroups: { [p: string]: string },
```
Comment (viceice) [inline lib/manager/regex/index.ts:43]:
```suggestion
  combinedGroups: Record<string, string>,
```

File: lib/manager/regex/index.ts [hunk 4]
```diff
@@ -108,18 +113,47 @@ function handleAny(
   return config.matchStrings
     .map((matchString) => regEx(matchString, 'g'))
     .flatMap((regex) => regexMatchAll(regex, content)) // match all regex to content, get all matches, reduce to single array
-    .map((matchResult) => createDependency(matchResult, config));
+    .map((matchResult) => createDependency(matchResult, null, config));
+}
+
+function mergeGroups(
+  mergedGroup: { [p: string]: string },
```
Comment (viceice) [inline lib/manager/regex/index.ts:120]: Use Record

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/regex/index.ts [hunk 5]
```diff
@@ -108,18 +113,47 @@ function handleAny(
   return config.matchStrings
     .map((matchString) => regEx(matchString, 'g'))
     .flatMap((regex) => regexMatchAll(regex, content)) // match all regex to content, get all matches, reduce to single array
-    .map((matchResult) => createDependency(matchResult, config));
+    .map((matchResult) => createDependency(matchResult, null, config));
+}
+
+function mergeGroups(
```
Comment (rarkins) [inline lib/manager/regex/index.ts:119]: Can you make sure there's no possibility for prototype pollution this way? :-/

Comment (secustor) [reply lib/manager/regex/index.ts:119]: I check now if the match group is a valid match field

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**